File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    void dfs(int row,int col,vector<vector<char>>& grid,vector<vector<int>>& vis,int n,int m) {
        vis[row][col]=1;
        int delrow[] = {1,0,-1,0};
        int delcol[] = {0,1,0,-1};
        for(int i=0;i<4;i++) {
            int nrow = row+delrow[i];
            int ncol = col + delcol[i];
            if(nrow>=0 && nrow<m && ncol>=0 && ncol<n && grid[nrow][ncol]=='1' && !vis[nrow][ncol]) {
                dfs(nrow,ncol,grid,vis,n,m);
            }
        }
    }
    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<int>> vis(m,vector<int>(n,0));
        int ans=0;
        for(int i=0;i<m;i++) {
            for(int j=0;j<n;j++) {
                if(!vis[i][j] && grid[i][j]=='1') {
                    dfs(i,j,grid,vis,n,m);
                    ans++;
                }
            }
        }
        return ans;
    }
};

File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int n = grid.length,m=grid[0].length;
        boolean [][] visited = new boolean[n][m];
        int islands = 0;
        for(int i=0;i<n;i++) {
            for(int j=0;j<m;j++) {
                if(!visited[i][j] && grid[i][j]=='1') {
                    dfs(i,j,grid,visited);
                    islands++;
                }
            }
        }
        return islands;
    }

    public void dfs(int r,int c, char[][] grid, boolean[][] visited) {
        int n = grid.length;
        int m = grid[0].length;

        if(r<0 || c<0 || r>=n || c >= m || grid[r][c] == '0' || visited[r][c]) {
            return;
        }
        visited[r][c] = true;
        dfs(r-1,c,grid,visited);
        dfs(r+1,c,grid,visited);
        dfs(r,c-1,grid,visited);
        dfs(r,c+1,grid,visited);
    }
}